error-codes/README.md

```diff
@@ -0,0 +1,97 @@
+# Error Codes
+
+## Introduction
+In the event that a node detects violation of a protocol or is unable to
+complete the necessary steps required for the protocol, it's useful to provide a
+reason for disconnection to the other end. This error code can be sent on both
+Connection Close and Stream Reset. Its purpose is similar to http response
+status. A server, on receiving an invalid request can reset the stream providing
+a `BAD_REQUEST` error code, when it's busy handling too many requests can
+provide a `RATE_LIMITED` error code, etc. An error code doesn't always indicate
+an error condition. For example, a connection may be closed prematurely because
+a connection to the same peer on a better transport is available. 
+
+## Semantics
+Error codes are unsigned 32bit integers. The range 0 to 10000 is reserved for
+libp2p errors. Application specific errors can be defined by protocols from
+integers outside of this range. Implementations supporting error codes MUST
+provide the error code provided by the other end to the application.
+
+Error codes provide a best effort guarantee that the error will be propogated to
+the applciation layer. This provides backwards compatibility with older nodes,
+allows smaller implementations and using transports that don't provide a
+mechanism to provide an error code. For example, Yamux has no equivalent of
+QUIC's [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5-4)
+frame that would tell the peer that the node has stopped reading. So there's no
+way of signaling an error while closing the read end of the stream on a yamux
+connection.
+
+### Connection Close and Stream Reset Error Codes
+Error codes are defined separately for Connection Close and Stream Reset. Stream
+Reset errors are from the range 0 to 5000 and Connection Close errors are from
+the range 5001 to 10000. Having separate errors for Connection Close and stream
+reset requires some overlap between the error code definitions but provides more
+information to the receiver. Receiving a `Bad Request: Connection Closed` error
+on reading from a stream also tells the receiver that no more streams can be
+started on the same connection. Implementations MUST provide the Connection
+Close error code on streams that are reset as a result of remote closing the
+connection. 
+
+For stream resets, when the underlying transport supports it, implementations
+SHOULD allow sending an error code on both closing the read side of the stream
+and resetting the write side of the stream. 
+
+## Libp2p Error Codes
+TODO!
```

Comment:
Is this section going to contain a list of canonical error codes that map to well defined error conditions?

I think that would be useful.